### PR TITLE
Reset password

### PR DESCRIPTION
**Pull request by jinroh**
This PR adds the possibility for the user to reset its password if forget. A token is generated and sent to the user via its mail.

The html templates could use some love from the front team (cc @gregorylegarec)

Some changes in the mail API have been introduced to better handle mail templates.

### CODE REVIEW

File: assets/templates/login.html [hunk 1]
```diff
@@ -58,6 +58,7 @@
               <div class="controls">
                 <button id="login-submit" form="login-form" type="submit">Log in</button>
               </div>
+              <a href="/auth/passphrase_reset">Forgot your password ?</a>
```
Comment (nono) [inline assets/templates/login.html]: No space before `?` in english

File: assets/templates/passphrase_renew.html [hunk 2]
```diff
@@ -0,0 +1,59 @@
+<!DOCTYPE html>
+<html lang="fr">
```
Comment (nono) [inline assets/templates/passphrase_renew.html]: We should probably stop hard-coding a `fr` lang in our templates

File: assets/templates/passphrase_reset.html [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+                <div class="svg-wrapper">
+                  <svg>
+                    <use xlink:href="#cozy-icon" />
+                  </svg>
+                </div>
+              </figure>
+            </header>
+            <div role="region">
+              <form id="reset-passphrase-form" method="POST" action="/auth/passphrase_reset" class="login auth">
+                <p class="help" id="login-password-tip">Are you sure you want to reset your password ?</p>
```
Comment (nono) [inline assets/templates/passphrase_reset.html]: Same here: no space before `?`

File: pkg/instance/instance.go [hunk 4]
```diff
@@ -534,15 +545,76 @@ func (i *Instance) RegisterPassphrase(pass, tok []byte) error {
...
+// renew its password.
+func (i *Instance) RequestPassphraseReset() error {
+	// If a registration token is set, we do not generate another token than the
+	// registration one, and bail.
+	if i.RegisterToken != nil {
+		return nil
+	}
+	// If a passphrase reset token is set and valid, we do not generate new one,
+	// and bail.
+	if i.PassphraseResetToken != nil &&
```
Comment (nono) [inline pkg/instance/instance.go]: If you have lost your passphrase and ask for a reset, you can only do it once for 24h. So, if you are unlucky and this mail is marked as spam, you have to wait 24h locked out of your cozy. Right? Maybe we can reduce the 24h delay to something a lot shorter, like 15 minutes. I'd also prefer if the passphrase reset token have a short life time for security reasons.

File: pkg/jobs/workers/mail_templates.go [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+	mailTemplater, err = newMailTemplater([]*MailTemplate{
+		{
+			Name:     "passphrase_reset",
+			BodyHTML: mailResetPassHTML,
+			BodyText: mailResetPassText,
+		},
+	})
+
+	if err != nil {
+		panic(err)
```
Comment (nono) [inline pkg/jobs/workers/mail_templates.go]: Maybe you can use [template.Must](https://golang.org/pkg/text/template/#Must) in newMailTemplater

File: web/auth/auth.go [hunk 6]
```diff
@@ -513,6 +520,75 @@ func checkRegistrationToken(next echo.HandlerFunc) echo.HandlerFunc {
...
+func passphraseReset(c echo.Context) error {
+	instance := middlewares.GetInstance(c)
+	// TODO: check user informations to allow the reset of the passphrase since
+	// this route is of course not protected by authentication/permission check.
+	if err := instance.RequestPassphraseReset(); err != nil {
+		return err
+	}
+	// Disconnect the user if it is logged in. The idea is that if the user
+	// (maybe by accident) asks for a passphrase reset while logged in, we log
+	// him out for him de re-go through the process of logging back-in. It is
```
Comment (nono) [inline web/auth/auth.go]: s/de/to/ ?

File: web/auth/auth.go [hunk 7]
```diff
@@ -513,6 +520,75 @@ func checkRegistrationToken(next echo.HandlerFunc) echo.HandlerFunc {
...
+	// more a UX choice than a "security" one.
+	if middlewares.IsLoggedIn(c) {
+		session, err := sessions.GetSession(c, instance)
+		if err == nil {
+			c.SetCookie(session.Delete(instance))
+		}
+	}
```
Comment (nono) [inline web/auth/auth.go]: Can you add an anti-CSRF on this route, like for `/auth/authorize`?

Comment (jinroh) [inline web/auth/auth.go]: Oh shoot, I thought it was "automatic" (and activated on all routes of `/auth`) but of course I need to add the token in the template and the routes.

Comment (jinroh) [inline web/auth/auth.go]: done 70e3f53e35a924b626c584a0949378563ad86838

File: pkg/instance/instance.go (hunk 4 above)
Comment (jinroh) [inline pkg/instance/instance.go]: Sharing your concerns, I reduced the token duration to 15 minutes.

**Comment by gregorylegarec** [comment]
Thanks @jinroh, we'll have a look on HTML templates ASAP.

**Pull request merged**